storage/innobase/btr/btr0cur.cc

```diff
@@ -1463,6 +1459,13 @@ dberr_t btr_cur_t::search_leaf(const dtuple_t *tuple, page_cur_mode_t mode,
     ut_ad(up_match != ULINT_UNDEFINED || mode != PAGE_CUR_LE);
     ut_ad(low_match != ULINT_UNDEFINED || mode != PAGE_CUR_LE);
 
+    if (latch_mode == BTR_MODIFY_TREE &&
+        btr_cur_need_opposite_intention(block->page, index()->is_clust(),
+                                        lock_intention,
+                                        node_ptr_max_size, compress_limit,
+                                        page_cur.rec))
+        goto need_opposite_intention;
```

Comment:
An interesting point to check should be that the code is working as expected in the newly moved position. It is likely there is not test to validate this functionality too. Otherwise we should already have failures.

Reply:
The `goto need_opposite_intention` is something that will escalate to exclusively acquiring `index->lock`, via `btr_cur_t::pessimistic_search_leaf()`. If we invoked that code path unnecessarily, there would be a performance regression due to increased contention on `index->lock`.

The cursor should be positioned correctly even if this fix is not present, but a later operation in the same mini-transaction could hang because it could violate the [WL#6326](https://dev.mysql.com/worklog/task/?id=6326) rules. The rules are supposed to be enforced here. Further code is only trying to catch misuse in some debug assertions (checking that we are holding an exclusive `index->lock` if we have to acquire new page latches up or back in the tree).

Reply:
yes, I understand it could be difficult to create a scenario. Perhaps I will give it a try later. The check looks correct to me in theory.
> because it could violate the [WL#6326](https://dev.mysql.com/worklog/task/?id=6326) rules.

Right. The WL is the best that we probably have for the rules. I remember there was some addition/correction needed there last time I was reviewing one bug with Yasufumi.